File: frontend/src/types/spotTrade.ts
/** Core SpotTrade interface matching backend SpotTradeRead schema. */
export interface SpotTrade {
  id: number
  trade_id: string
  portfolio_id: number | null
  counterparty_id: number | null
  portfolio_name: string | null
  counterparty_name: string | null

  // Trade details
  ccy_pair: string | null
  direction: string | null      // 买入 / 卖出
  event_type: string | null     // 正常 / 期权行权衍生

  deal_price: number | null
  ccy1_amount: number | null
  ccy2_amount: number | null
  ccy1: string | null
  ccy2: string | null

  // Dates
  trade_date: string | null
  settlement_date: string | null

  // Status and meta
  source: string | null
  venue: string | null

  // Timestamps
  created_at: string | null
  updated_at: string | null
}

export interface SpotTradeListResponse {
  data: SpotTrade[]
  total: number
  page: number
  page_size: number
}

export interface SpotTradeFilterParams {
  page?: number
  page_size?: number
  sort_by?: string
  sort_order?: string
  portfolio_id?: number | null
  counterparty_id?: number | null
  ccy_pair?: string
  direction?: string
  event_type?: string
  trade_date_from?: string
  trade_date_to?: string
  settlement_date_from?: string
  settlement_date_to?: string
  search?: string
}

export interface SpotTradeCreate {
  trade_id: string
  portfolio_id?: number | null
  portfolio_name?: string | null
  counterparty_name?: string | null
  ccy_pair?: string | null
  direction?: string | null
  event_type?: string | null
  deal_price?: number | null
  ccy1_amount?: number | null
  ccy2_amount?: number | null
  trade_date?: string | null
  settlement_date?: string | null
  source?: string | null
  venue?: string | null
}

export interface SpotTradeUpdate {
  trade_id?: string | null
  portfolio_id?: number | null
  portfolio_name?: string | null
  counterparty_name?: string | null
  ccy_pair?: string | null
  direction?: string | null
  event_type?: string | null
  deal_price?: number | null
  ccy1_amount?: number | null
  ccy2_amount?: number | null
  trade_date?: string | null
  settlement_date?: string | null
  source?: string | null
  venue?: string | null
}

File: backend/app/schemas/spot_trade.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict


class SpotTradeBase(BaseModel):
    """Common fields shared across spot trade schemas."""

    trade_id: str | None = None
    portfolio_id: int | None = None
    counterparty_id: int | None = None
    portfolio_name: str | None = None
    counterparty_name: str | None = None

    # Trade details
    ccy_pair: str | None = None
    direction: str | None = None
    event_type: str | None = None

    deal_price: float | None = None
    ccy1_amount: float | None = None
    ccy2_amount: float | None = None
    ccy1: str | None = None
    ccy2: str | None = None

    # Dates
    trade_date: date | None = None
    settlement_date: date | None = None

    # Status and meta
    source: str | None = None
    venue: str | None = None


class SpotTradeRead(SpotTradeBase):
    """Schema for reading a SpotTrade (API response)."""

    model_config = ConfigDict(from_attributes=True)

    id: int
    trade_id: str
    created_at: datetime | None = None
    updated_at: datetime | None = None


class SpotTradeCreate(SpotTradeBase):
    """Schema for creating a new spot trade manually."""

    trade_id: str


class SpotTradeUpdate(SpotTradeBase):
    """Fields that can be updated on a spot trade."""

    pass


class SpotTradeListResponse(BaseModel):
    """Paginated list response for spot trades."""

    data: list[SpotTradeRead]
    total: int
    page: int
    page_size: int


class SpotTradeFilterParams(BaseModel):
    """Query parameters for filtering spot trades list."""

    page: int = 1
    page_size: int = 50
    sort_by: str = "trade_date"
    sort_order: str = "desc"
    portfolio_id: int | None = None
    counterparty_id: int | None = None
    ccy_pair: str | None = None
    direction: str | None = None
    event_type: str | None = None
    trade_date_from: date | None = None
    trade_date_to: date | None = None
    settlement_date_from: date | None = None
    settlement_date_to: date | None = None
    search: str | None = None
